src/ray/object_manager/transfer_queue.h

```diff
@@ -0,0 +1,115 @@
+#ifndef RAY_TRANSFER_QUEUE_H
+#define RAY_TRANSFER_QUEUE_H
+
+#include <algorithm>
+#include <cstdint>
+#include <deque>
+#include <map>
+#include <memory>
+#include <thread>
+
+#include <boost/asio.hpp>
+#include <boost/asio/error.hpp>
+#include <boost/bind.hpp>
+
+#include "ray/id.h"
+#include "ray/status.h"
+
+#include "format/object_manager_generated.h"
+#include "object_directory.h"
+#include "object_manager_client_connection.h"
+
+namespace asio = boost::asio;
+
+namespace ray {
+
+class TransferQueue {
+ public:
+  enum TransferType { SEND = 1, RECEIVE };
+
+  /// Context maintained during an object send.
+  struct SendContext {
+    ClientID client_id;
+    ObjectID object_id;
+    int64_t object_size;
+    uint8_t *data;
+  };
+
+  /// The structure used in the send queue.
+  struct SendRequest {
+    ClientID client_id;
+    ObjectID object_id;
+    friend bool operator==(const SendRequest &o1, const SendRequest &o2) {
```

Comment:
Is this necessary? I would rather we just did this in the .cc file than defining it here.

Reply:
Yes, it's necessary.

Reply:
Can you explain why?

Reply:
@elibol , see an example of `bool operator==(const ResourceSet &rhs) const;` here:
https://github.com/ray-project/ray/blob/xray-task-forwarding/src/ray/raylet/scheduling_resources.h#L44

Reply:
The equality operator is needed to avoid adding duplicate entries in `send_queue_`.